src/librustc_middle/ty/sty.rs

```diff
@@ -319,19 +319,42 @@ pub struct ClosureSubsts<'tcx> {
     pub substs: SubstsRef<'tcx>,
 }
 
-/// Struct returned by `split()`. Note that these are subslices of the
-/// parent slice and not canonical substs themselves.
-struct SplitClosureSubsts<'tcx> {
-    closure_kind_ty: GenericArg<'tcx>,
-    closure_sig_as_fn_ptr_ty: GenericArg<'tcx>,
-    tupled_upvars_ty: GenericArg<'tcx>,
+/// Struct returned by `split()`.
+// FIXME(eddyb) find a better name than `Split...` - perhaps `ClosureSubstsParts`?
+pub struct SplitClosureSubsts<T> {
+    pub closure_kind_ty: T,
+    pub closure_sig_as_fn_ptr_ty: T,
+    pub tupled_upvars_ty: T,
 }
 
 impl<'tcx> ClosureSubsts<'tcx> {
-    /// Divides the closure substs into their respective
-    /// components. Single source of truth with respect to the
-    /// ordering.
-    fn split(self) -> SplitClosureSubsts<'tcx> {
+    /// Construct `ClosureSubsts` from base `Substs` (i.e. of the parent function)
+    /// and `SplitClosureSubsts` containing the additional closure-specific components.
+    pub fn new(
+        tcx: TyCtxt<'tcx>,
+        base_substs: SubstsRef<'tcx>,
+        parts: SplitClosureSubsts<Ty<'tcx>>,
+    ) -> ClosureSubsts<'tcx> {
+        ClosureSubsts {
+            substs: tcx.mk_substs(
+                base_substs.iter().chain(
+                    [parts.closure_kind_ty, parts.closure_sig_as_fn_ptr_ty, parts.tupled_upvars_ty]
+                        .iter()
+                        .map(|&ty| ty.into()),
+                ),
+            ),
+        }
+    }
+
+    /// Returns the non-closure-specific base `Substs` (i.e. of the parent function).
+    /// The ordering assumed here must match that used by `ClosureSubsts::new` above.
+    pub fn base_substs(self) -> &'tcx [GenericArg<'tcx>] {
+        &self.substs[..self.substs.len() - 3]
```

Comment:
Should this assert `is_valid(self)`? Someone might copy paste some code and end up calling [`as_generator`](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/ty/subst.rs.html#201-203) and try read the base_substs without checking is_valid and this would return incorrect information. 

Might save someone fair amount of debugging time.